src/tangle/sources/inx.rs

```diff
@@ -1,2 +1,90 @@
 // Copyright 2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use futures::{stream::BoxStream, StreamExt, TryStreamExt};
+
+use super::{BlockData, InputSource, MilestoneData};
+use crate::{
+    inx::{Inx, InxError, MarkerMessage},
+    tangle::{ledger_updates::LedgerUpdateStore, milestone_range::MilestoneRange},
+    types::{
+        ledger::{LedgerSpent, MilestoneIndexTimestamp},
+        tangle::MilestoneIndex,
+    },
+};
+
+#[async_trait]
+impl InputSource for Inx {
+    type Error = InxError;
+
+    async fn milestone_stream(
+        &self,
+        range: MilestoneRange,
+    ) -> Result<BoxStream<Result<MilestoneData, Self::Error>>, Self::Error> {
+        let mut inx = self.clone();
+        Ok(Box::pin(
+            inx.listen_to_confirmed_milestones(range.into())
+                .await?
+                .and_then(|msg| async move {
+                    let payload = if let iota_types::block::payload::Payload::Milestone(payload) =
+                        msg.milestone.milestone.inner_unverified()?
+                    {
+                        payload.into()
+                    } else {
+                        // The raw data is guaranteed to contain a milestone payload.
+                        unreachable!();
```

Comment:
I think I prefer putting the documentation into the panic like so:
```suggestion
                        unreachable!("Raw milestone data has to contain a milestone payload");
```